### ISSUE TITLE
[feature req] add OIDC authentication + roles

### ISSUE THREAD

**Issue by andrewzah** [navidrome/navidrome#858]
https://swagger.io/docs/specification/authentication/openid-connect-discovery/

This would allow things like SSO through an external service like Keycloak, which generates `.well-known` automatically. A generic OIDC authenticator is preferable to service-specific logins like Google, Facebook, etc.

One way to do this would be to have multiple Authenticators, like an `OIDCAuthenticator` and a `LocalAuthenticator`, which is the current behavior. Having an option to disable the `LocalAuthenticator` entirely and only populate users through the `OIDCAuthenticator` would be preferable.

For admins: One solution is either the `access` token or the `auth` token could have a `roles` object, which would look for a role named `admin`.

**Comment by deluan** [comment]
Multiple authentication backend will be implemented at some point, but the priority now is to decide how to work around the Subsonic API limitations regarding authentication: https://github.com/navidrome/navidrome/issues/202#issuecomment-797640193.

**Comment by rmccullagh99** [comment]
Bumping this as something I'm still keen to see in Navidrome.
Understand this is waiting for the revisions to the API, but very much looking forward to this.

**Comment by Mihara** [comment]
This bot is a bit overeager, the issue is still very much relevant.

**Comment by phw** [comment]
> Thanks! Multiple authentication backend will be implemented at some point, but the priority now is to decide how to work around the Subsonic API limitations regarding authentication: [#202 (comment)](https://github.com/navidrome/navidrome/issues/202#issuecomment-797640193).

One way to handle this would be to provide separate Subsonic authentication passwords. A password could be autogenerated when a user gets created. While this password will be stored in database it would only be used for subsonic authentication and would not allow login into the UI or access anything not supported by subsonic.

The main user password could then be properly hashed.

This does not fully eliminate all risks, but limits the impact of the password. Especially for admin accounts, but also normal accounts.

**Comment by andrewzah** [comment]
Agreed, generating a random token for subsonic auth seems like a reasonable way to address this.

**Comment by onedr0p** [comment]
This bot is a bit overeager, the issue is still very much relevant.

**Comment by HearthCore** [comment]
Indeed, SSO via OIDC is thanks to Authentik and Authelia easily embeddable.

**Comment by DDriggs00** [comment]
One way this could work is using app passwords for subsonic, like nextcloud uses for Caldav when it is using OIDC.

**Comment by ForsakenRei** [comment]
Looking forward for OIDC support with Authentik, too

**Comment by DDriggs00** [comment]
The current SSO method is incompatible with the subsonic API. This issue is still very relevant.

**Comment by andrewzah** [comment]
Yep, still relevant.

**Comment by scarpentier** [comment]
Bumping so it does not become stale. Is this on the roadmap at least? Willing to put a bounty on this if anyone's interested.

**Comment by cyndane31** [comment]
> Bumping so it does not become stale. Is this on the roadmap at least? Willing to put a bounty on this if anyone's interested.

I would contribute to said bounty as well!

**Comment by djthread** [comment]
I would adore the ability to hook my Pocket ID up to Navidrome.

Much appreciation for the hard work on Navidrome - I love this tool.

**Comment by homelab-hobbyist** [comment]
Another vote for OIDC support.

**Comment by stavros-k** [comment]
@deluan  Please re-open/clone this as OP closed it..

**Comment by penchochris** [comment]
+1 here

**Comment by anony253** [comment]
+1, can add to bounty as well

**Comment by booinspace** [comment]
Still relevant! OIDC support would be beautiful

**Comment by raudraido** [comment]
+1

**Comment by ildus** [comment]
+1

**Comment by modzilla99** [comment]
Yes, OIDC would be pretty neat!

**Comment by Akruidenberg** [comment]
+1 for me!

**Comment by KucharczykL** [comment]
I would also like OIDC support.

**Comment by nadir-ishiguro** [comment]
Just for those people not familiar with software development:

This is not a thread. It's a github issue. 

People are subscribed to it in case anything changes on the developer side.

Commenting with "+1" or "me too" comments pings everybody else and will if anything harden any bias the developers might have against implementing the feature.

Learning how to implement such a feature in a way that fits well into the project and seeking the input and understanding of the development team is the only chance of ever getting this done. 

If you yourself can't do that and are not willing to learn, the next best thing is finding people that do and encouraging and supporting them, not nagging them.

This kind of endless thread is what leads to people stopping contributing to open source projects.

**Comment by KucharczykL** [comment]
I also dislike people endlessly saying "+1", but I feel like there's an important nuance to it: you need a place that people can watch for updates.

This issue has been locked so people feel like they need to say something otherwise the developers might think this feature is not important. Also, this repository has made it so that open topics are automatically closed if no comments are posted for a while. This has conditioned users to feel like posting "I also want this" is OK.

It could alternatively be converted to a discussion here on GitHub.
